### ISSUE TITLE
Random number generation

### ISSUE THREAD

**Issue by NeilGirdhar** [data-apis/array-api-extra#308]
I'd love to have my [Efax](https://github.com/NeilGirdhar/efax) library entirely in the Array API.  The only thing missing now is:
* random number generation,
* JIT,
* automatic gradients, and
* Jax's "Pytree" system.

Random number generation is probably the biggest blocker though.  I think that the SciPy people intended to eventually add some kind of generic random number generation.  Is this the sort of thing that might find its way into this project?  Seems like SciPy is the wrong home for it since they don't even intend to populate SciPy with all of the distributions they would need.

**Comment by lucascolley** [comment]
@mdhaber I wonder whether you have any thoughts on this

**Comment by NeilGirdhar** [comment]
> check out [data-apis/array-api#431](https://github.com/data-apis/array-api/issues/431) if you haven't already, btw

I have seen it.  Regarding the issues with Jax, I'm not sure they're much of a blocker.  Consider an interface like this:
```python
class RngStream:
    def __init__(self, key: KeyArray, count: JaxIntegralArray | None = None) -> None:
        super().__init__()
        if count is None:
            count = jnp.zeros((), dtype=jnp.uint32)
        else:
            assert count.ndim == 0
        self._key = key
        self._count = count

    def key(self) -> KeyArray:
        key = jr.fold_in(self._key, self._count)
        self._count += 1
        return key

    def fork(self, samples: int) -> KeyArray:
        return jr.split(self.key(), samples)

def uniform(stream: RngStream, ...) -> ...:
  ...  # etc.
```

This should be good enough for writing generic algorithms, and works fine for Jax provided you only pass `KeyArray` objects to jitted functions, which is not the responsibility of `xpx`.

**Comment by izaid** [comment]
Yes please!

**Comment by NeilGirdhar** [comment]
[Here](https://github.com/data-apis/array-api-extra/pull/326)'s a rough draft of how I imagine this working.  This should be good enough to test a statistical library like SciPy's and might be good enough to use in a machine learning project.  Comments welcome.

**Comment by crusaderky** [comment]
> [Here](https://github.com/data-apis/array-api-extra/pull/326)'s a rough draft of how I imagine this working. This should be good enough to test a statistical library like SciPy's and might be good enough to use in a machine learning project. Comments welcome.

Sorry I must be dense - could you show a mock-up of that working with a generic function `f(rng: Generator, xp: Namespace) -> Array`, which may or may not be wrapped with`jax.jit`? I can't understand. I assume your POC is missing the pytree conversion bit?

**Comment by lucascolley** [comment]
My main high-level concern is adding something to the API which does not have a generic solution. I think there is definitely a place for this work, which can fallback to NumPy and device transfers for unknown libraries, but I'm not 100% sure if xpx is the right place.

This is by no means a dealbreaker, I just want to think through the implications for library scope. ref https://data-apis.org/array-api-extra/#scope

**Comment by NeilGirdhar** [comment]
> My main high-level concern is adding something to the API which does not have a generic solution.

Could you elaborate on how this isn't a "generic solution"?  I'm not sure I understand.  From what I can see, `Generator` could work with at least Numpy, Jax, and Pytorch.  I don't know the other backends though.

**Comment by lucascolley** [comment]
Library X, where all we know about library X is that it implements the standard.

**Comment by lucascolley** [comment]
I think the main reason for that thought is that we do not want to introduce a dependency on an array library.

The other option is to just make it work for libraries which we choose to support, and error out otherwise. But I am a little worried that that opens a can of worms for:
- how many / which libraries we should choose
- other features which cannot be implemented for generic libraries, but only for a few

Practically, in the short term, I don't think these are huge problems. But I think they definitely could be problematic long-term.

**Comment by NeilGirdhar** [comment]
> The other option is to just make it work for libraries which we choose to support, and error out otherwise. 

Right, that's what I thought we would do.  I don't think there's any way around that since there's no random number generation in the Array API.

> * how many / which libraries we should choose

As many as possible?

> * other features which cannot be implemented for generic libraries, but only for a few

Generally, they don't have all the same random generation functions so there may be limits.  However, we may actually be able to have a roughly universal interface.  For example, Pytorch can't sample from the Gumbel distribution, but Jax can.  We could use the standard uniform sampler with some additional code to fill in the missing function for all libraries that are missing Gumbel sampling.

What do you think?

**Comment by lucascolley** [comment]
> - other features which cannot be implemented for generic libraries, but only for a few
> 
> Generally, they don't have all the same random generation functions so there may be limits

I'm speaking generally about library scope here, not just rng

**Comment by lucascolley** [comment]
> @lucascolley and @crusaderky are provide excellent and fast reviews

Great to hear :) although probably worth noting that all of my work here is on a volunteer basis, I can't guarantee how much time I'll have to give to the project going forwards (no plans of abandoning it though, don't worry!)

> Why don't we consider expanding the scope of xpx to include functions that are not just compositions of Array API functions on a case-by-case basis?

We definitely can consider it. I just want to make sure that it is considered well.

**Comment by NeilGirdhar** [comment]
@34j Yes that makes sense.  We could broaden the annotation for `create_generator` and add namespace-specific logic.

(I won't have time to work on this for quite a while as I'm on vacation.)

**Comment by 34j** [comment]
In addition, I would like there to be a discussion on how much method and argument names for Generator should be compliant with Numpy or other libraries.

Personally, I feel Numpy's generator is confusing because the method for uniform random is random() instead of uniform (), and the return shape is specified by size instead of shape.

**Comment by 34j** [comment]
@ntessore 
> where `xpx.jit(f)` means:
> 
>     * Use `jax.tree.map` to transform every `JaxGenerator` into a special `WrappedKeyArray`.
> 
>     * Apply `jax.jit` to `f` and call it.
> 
>     * Use`jax.tree.map` to transform every `WrappedKeyArray` back into a `JaxGenerator`.
> 
> 
> We can start another issue for generalizing the jit decorator in this way if you like? If we go down this route, we'll also need to generalize the other traced functions: `vjp`, `jvp`, `grad`, `vmap`, etc.

To avoid introducing "wrapped" decorators like `xpx.jit`, first and third item need to be dropped.

- To drop the first, as you suggested, `Generator.fork()` need to return an object natively supported by `jit`.
- To drop the 3rd, the key needs to be converted to Generator inside the function. Instead of introducing a generic function as you suggested, I would suggest to simply forbid the function to take `Generator` as an argument and always call `xpx.generator` within a function (or maybe we could add decorator to do this automatically?)

> def uniform_power(shape: tuple[int, ...], xp: ArrayNamespace, key_or_seed_or_none: Any = None, dtype: Any, device: Any) -> Array:
>     generator = xpx.generator(xp, key_or_seed_or_none)
>     return generator.uniform(shape=shape, dtype=dtype, device=device) ** 2

**Comment by lucascolley** [comment]
> The whole point of this library is to be able to write backend-independent tests of Array API code.

Keep in mind that this is not the only use-case. Parts of `scipy.stats` which accept NumPy generators, for example, would be interested in including this in production code, not just tests.

**Comment by lucascolley** [comment]
Yes. I would suggest that one of you (@34j @NeilGirdhar @ntessore) makes a new repo and invites the others as collaborators, where you can work through an issue like https://github.com/data-apis/array-api-typing/issues/22. If there is rough agreement on what should be worked on and a prototype going, then it might be time to create data-apis/array-api-rng.

**Comment by NeilGirdhar** [comment]
Done: https://github.com/NeilGirdhar/array-api-rng/ I'm on vacation for the next 10 days, but I added the collaborators.

**Comment by 34j** [comment]
@NeilGirdhar Any progress? I'm not inclined to lift the setup *on my own* because I don't understand their setup.

**Comment by 34j** [comment]
> [@34j](https://github.com/34j) You're welcome to set up the repo however you think is best. I just thought we should steal whatever ideas from their repo since I like some of the choices they made: Ruff, clean pyproject.toml, etc.

Thanks for your welcome but I won't be able to help for a while, sorry.

**Comment by lucascolley** [comment]
going to convert to a discussion — at least for now, array-api-extra is not the right place for this to be developed. Let's keep pointing to https://github.com/NeilGirdhar/array-api-rng/issues for now until we are roughly converging on a prototype.
